File: static/js/dr/verify-mirror.js
// static/js/dr/verify-mirror.js
// ADR-008 §4.2: кросс-девайсный мирор OOB-верификаций участников.
//
// Проблема: OOB-верификация («я сверил account-Ed пира вне канала») хранится
// локально (vortex_verified_ed_<peer>) → не переживает переход на другое устройство.
// Решение, НЕ ломающее «сервер недоверен»: устройство подписывает атестацию своим
// DEVICE-ключом (account-Ed приватного на линкованном устройстве нет — blast-radius),
// приложив device-cert (account-Ed подписал device-триплет). Сервер хранит и отдаёт
// владельцу его же записи, но подделать не может (нет приватных). Другое устройство
// владельца верифицирует device-cert→СВОЙ account-Ed + attest-sig и применяет.
//
// Мирор — лишь синхронизация локального hint. Заворачивание room-key (G3) НЕЗАВИСИМО
// ре-чекает живой ed против сохранённого верифицированного → подделка/откат мирора не
// даёт завернуть на подменённый ключ. Дормантно за флагом vortex_verify_mirror_enabled
// (дефолт ВЫКЛ, opt-in). Kill-switch — снять флаг.

import { edSign, edVerify, loadAccountLinkMaterial, loadEd25519Identity } from './prekeys.js';
import { loadOrCreateDeviceIdentity, verifyDeviceCert } from './device-identity.js';
import { markIdentityVerified, unmarkIdentityVerified } from './member-verify.js';
import { api } from '../utils.js';

function _mirrorEnabled() {
    try { return localStorage.getItem('vortex_verify_mirror_enabled') === '1'; } catch { return false; }
}

/**
 * Каноническая полезная нагрузка атестации (cross-impl ЗАФИКСИРОВАНА, §4.7 —
 * sort/регистр не при чём, но ed в LOWERCASE, разделитель ':').
 *   "vortex-verify-attest:v1:" + owner + ":" + peer + ":" + ed_lower + ":" + state + ":" + signed_at
 * @returns {Uint8Array}
 */
export function attestPayload(ownerUserId, peerUserId, verifiedEdHex, state, signedAt) {
    const s = `vortex-verify-attest:v1:${ownerUserId}:${peerUserId}:${String(verifiedEdHex).toLowerCase()}:${state}:${signedAt}`;
    return new TextEncoder().encode(s);
}

/** account-Ed pub (hex) этого аккаунта, из любого доступного источника. */
function _myAccountEdPub(userId) {
    return loadAccountLinkMaterial(userId)?.accountEdPub
        || loadEd25519Identity()?.pubHex
        || null;
}

/**
 * Подписать и опубликовать атестацию верификации (или её отзыв). No-op при
 * выключенном флаге или отсутствии device-cert (устройство ещё не авторизовано).
 * @param {number|string} peerUserId
 * @param {string} verifiedEdHex — account-Ed пира, который я сверил
 * @param {'verified'|'revoked'} state
 */
export async function publishAttestation(peerUserId, verifiedEdHex, state) {
    if (!_mirrorEnabled()) return;
    const owner = window.AppState?.user?.user_id;
    if (!owner || !verifiedEdHex) return;
    try {
        const dev = await loadOrCreateDeviceIdentity();
        if (!dev?.certSig || !dev.signPriv) return;      // cert не выпущен → нечем подтвердить авторство
        const signedAt = Math.floor(Date.now() / 1000);
        const attest_sig = await edSign(dev.signPriv, attestPayload(owner, peerUserId, verifiedEdHex, state, signedAt));
        await api('POST', '/api/verify/attestations', {
            peer_user_id:     Number(peerUserId),
            verified_ed:      String(verifiedEdHex).toLowerCase(),
            state,
            signed_at:        signedAt,
            client_device_id: dev.deviceId,
            device_x3dh_pub:  dev.x3dhPub,
            device_sign_pub:  dev.signPub,
            device_cert_sig:  dev.certSig,
            attest_sig,
        });
    } catch (e) {
        console.debug('[verify-mirror] publish skipped:', e?.message);
    }
}

/**
 * Синхронизировать OOB-верификации с сервера на это устройство. Для каждой записи:
 * device-cert должен вести к МОЕМУ account-Ed (подписант — авторизованное устройство
 * этого аккаунта) И attest-sig валиден под device-ключом. Только тогда применяем.
 * @returns {Promise<{applied:number, rejected:number}>}
 */
export async function syncAttestations() {
    const out = { applied: 0, rejected: 0 };
    if (!_mirrorEnabled()) return out;
    const owner = window.AppState?.user?.user_id;
    const myAccountEd = owner ? _myAccountEdPub(owner) : null;
    if (!owner || !myAccountEd) return out;

    let rows = [];
    try {
        const resp = await api('GET', '/api/verify/attestations');
        rows = resp?.attestations || [];
    } catch (e) {
        console.debug('[verify-mirror] sync fetch failed:', e?.message);
        return out;
    }

    for (const r of rows) {
        try {
            // 1. Подписант — авторизованное устройство МОЕГО аккаунта (cert→мой account-Ed).
            const certOk = await verifyDeviceCert(
                r.client_device_id, r.device_x3dh_pub, r.device_sign_pub, r.device_cert_sig, myAccountEd);
            if (!certOk) { out.rejected++; continue; }
            // 2. Атестацию реально подписал этот device-ключ.
            const sigOk = await edVerify(
                r.device_sign_pub,
                attestPayload(owner, r.peer_user_id, r.verified_ed, r.state, r.signed_at),
                r.attest_sig);
            if (!sigOk) { out.rejected++; continue; }

            if (r.state === 'verified') markIdentityVerified(r.peer_user_id, String(r.verified_ed).toLowerCase());
            else if (r.state === 'revoked') unmarkIdentityVerified(r.peer_user_id);
            out.applied++;
        } catch { out.rejected++; }
    }
    return out;
}

File: app/security/verify_mirror.py
"""app/security/verify_mirror.py — кросс-девайсный мирор OOB-верификаций (ADR-008 §4.2).

Сервер здесь — НЕподделываемое хранилище, не источник доверия. Он не имеет ни
device-, ни account-приватного ключа владельца → сфабриковать «я сверил пира» не
может. Каждая запись подписана device-ключом владельца, а device-cert связывает
этот ключ с account-Ed владельца. Верифицирует и применяет запись ДРУГОЕ устройство
владельца (device-cert→свой account-Ed + attest-sig). Сервер лишь хранит и отдаёт
владельцу его же записи (owner == current_user; чужие недоступны).

Заворачивание room-key (ADR-008 G3) всё равно НЕЗАВИСИМО ре-чекает живой ed против
сохранённого верифицированного — мирор синхронизирует лишь локальный hint, не
подменяет проверку. Компрометация сервера не даёт заворачивания на подменённый ключ.
"""

import re

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import User, VerificationAttestation
from app.security.auth_jwt import get_current_user

router = APIRouter(prefix="/api/verify", tags=["verify-mirror"])

_HEX64 = re.compile(r"^[0-9a-fA-F]{64}$")
_HEX128 = re.compile(r"^[0-9a-fA-F]{128}$")
_HEX32 = re.compile(r"^[0-9a-fA-F]{32}$")


class AttestationIn(BaseModel):
    peer_user_id: int
    verified_ed: str
    state: str  # verified | revoked
    signed_at: int
    client_device_id: str
    device_x3dh_pub: str
    device_sign_pub: str
    device_cert_sig: str
    attest_sig: str


@router.post("/attestations")
async def put_attestation(
    body: AttestationIn,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Сохранить подписанную атестацию верификации (latest-per-peer выигрывает).

    Гигиена формата (не проверка подписи — это работа устройства-потребителя):
    сервер отклоняет мусор по длинам/hex, но НЕ верифицирует крипто и не может.
    Монотонность по signed_at не даёт откатить свежую запись реплеем старой.
    """
    if body.state not in ("verified", "revoked"):
        raise HTTPException(400, "state must be verified|revoked")
    if (
        not _HEX64.match(body.verified_ed)
        or not _HEX64.match(body.device_x3dh_pub)
        or not _HEX64.match(body.device_sign_pub)
    ):
        raise HTTPException(400, "bad hex-64 field")
    if not _HEX128.match(body.device_cert_sig) or not _HEX128.match(body.attest_sig):
        raise HTTPException(400, "bad hex-128 signature")
    if not _HEX32.match(body.client_device_id):
        raise HTTPException(400, "bad client_device_id")

    existing = (
        db.query(VerificationAttestation)
        .filter(
            VerificationAttestation.owner_user_id == user.id,
            VerificationAttestation.peer_user_id == body.peer_user_id,
        )
        .first()
    )
    # Реплей старой записи не должен откатывать свежую (rollback-guard).
    if existing and existing.signed_at >= body.signed_at:
        return {"ok": True, "applied": False}

    if existing:
        existing.verified_ed = body.verified_ed.lower()
        existing.state = body.state
        existing.signed_at = body.signed_at
        existing.client_device_id = body.client_device_id.lower()
        existing.device_x3dh_pub = body.device_x3dh_pub.lower()
        existing.device_sign_pub = body.device_sign_pub.lower()
        existing.device_cert_sig = body.device_cert_sig.lower()
        existing.attest_sig = body.attest_sig.lower()
    else:
        db.add(
            VerificationAttestation(
                owner_user_id=user.id,
                peer_user_id=body.peer_user_id,
                verified_ed=body.verified_ed.lower(),
                state=body.state,
                signed_at=body.signed_at,
                client_device_id=body.client_device_id.lower(),
                device_x3dh_pub=body.device_x3dh_pub.lower(),
                device_sign_pub=body.device_sign_pub.lower(),
                device_cert_sig=body.device_cert_sig.lower(),
                attest_sig=body.attest_sig.lower(),
            )
        )
    db.commit()
    return {"ok": True, "applied": True}


@router.get("/attestations")
async def get_attestations(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Отдать владельцу ЕГО ЖЕ атестации (latest-per-peer). Чужие недоступны."""
    rows = db.query(VerificationAttestation).filter(VerificationAttestation.owner_user_id == user.id).all()
    return {
        "attestations": [
            {
                "peer_user_id": r.peer_user_id,
                "verified_ed": r.verified_ed,
                "state": r.state,
                "signed_at": r.signed_at,
                "client_device_id": r.client_device_id,
                "device_x3dh_pub": r.device_x3dh_pub,
                "device_sign_pub": r.device_sign_pub,
                "device_cert_sig": r.device_cert_sig,
                "attest_sig": r.attest_sig,
            }
            for r in rows
        ],
    }
